### PR TITLE
Smart hint and text box style refactor

### PR DESCRIPTION
**Pull request by nicolaihenriksen**
This PR builds on top of the "RFC" around the `TextBox` style change: #3461 

### (Visual) Breaking changes !!!
This PR includes breaking changes.
* Applying `wpf:HintAssist.IsFloating="True"` is no longer enough to change from the default style to the floating hint style. Explicit opt-in in the form of `Style="{StaticResource MaterialDesignFloatingHintTextBox}"` is now required.
  * Although this is a breaking API change, I believe it is a good change. It allows better control to override properties because they are now set in the style rather than in a trigger inside the `ControlTemplate.Triggers` collection.
* Minor changes to the `TextBox` control height in the various styles are to be expected; this also affects some of the UI tests which have been temporarily modified to run green. When all styles adopt the new `SmartHint` approach, they should be able to run "normally" again.

### General idea
image
Instead of a `SmartHint` consumer relying on a lot of "magic numbers" to place the floating hint correctly, this PR attempts to let the `SmartHint` handle the placement of the floating hint itself, with a few supporting properties.

In the layout above, the green box represents the `PART_ContentHost` of the `TextBox` style which holds the user input. The red box represents the `SmartHint` as a whole; this element spans the width of the prefix-text, the user input, and the suffix-text in the layout to be able to place float the hint to either align with prefix/suffix, or with the user input. The blue/purple box represents the "scaled hint" inside of the `SmartHint`.

From a `SmartHint` consumers point-of-view (e.g. the `TextBox` style), the idea is that you provide a `SmartHint.FloatingTarget`, which similarly to the `Popup.PlacementTarget` is a dependency property where the caller provides a target element which the hint will then (vertically) float on top of. In the example above, the `SmartHint.FloatingTarget` is set to `PART_ContentHost` to always float the hint directly on top of the user input. There are a few other supporting properties to enable the hint to horizontally align with either the user input or the prefix/suffix texts; depending on the value of `HorizontalContentAlignment` as well.

### Examples
| Align with prefix | Align with user input |
| --- | --- |
| NonOutlineAllignWithPrefix | NonOutlineAllignWithUserInput |
| OutlineAllignWithPrefix | OutlineAllignWithInput |

In the - slowed down - animations above, you should note that the whole `SmartHint` (red box) floats upwards, but only far enough such that the "scaled down hint" (blue/purple box) rests directly on top of the `SmartHint.FloatingTarget`. Effectively overlapping the user input with the "empty space" left after scaling the hint down. In the outlined style, the `SmartHint.FloatingTarget` is set as the outer `Border`, and a few supporting properties have been set to allow the hint to center align (vertically) with the border line.

This approach, in my opinion, works better than the old implementation. The intention is to retain all existing functionality (but improve upon it), and also add support for change of font size and multi-line input. I believe the branch in its current state nearly has that, with just a few minor things I need to look into / clean up.

The intention is to use the `TextBox` style as the guinea pig for adopting the new approach, and then to roll it out to all the other styles containing a `SmartHint` afterwards. Backwards compatibility is intended to be maintained, so if a non-default value for `HintAssist.FloatingOffset` is supplied, that will still be respected. It could however be considered a both a behavioral breaking change (alignment with prefix/suffix/user-input) and a (slight) visual breaking change.

### Funkyness
image
This particular (existing) feature is rather funky. I retained back-compat of it, but it is not aligned across the styles at all. It seems like a shortcut instead of doing an actual MD3 version of the styles. It basically changes the default `TextBox` style so that the underline behaves slightly different on hover.

Any comments/feedback/pushback is welcome.

### CODE REVIEW

File: MaterialDesignThemes.Wpf/Converters/BooleanToDashStyleConverter.cs [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+public class BooleanToDashStyleConverter : IValueConverter
+{
+    public DashStyle? TrueValue { get; set; }
+    public DashStyle? FalseValue { get; set; }
+
+    public object? Convert(object? value, Type targetType, object? parameter, CultureInfo culture)
+        => value is true ? TrueValue : FalseValue;
+
+    public object? ConvertBack(object? value, Type targetType, object? parameter, CultureInfo culture)
+        => throw new NotImplementedException();
```
[the remark this answers is not in the document]
Comment (nicolaihenriksen) [reply MaterialDesignThemes.Wpf/Converters/BooleanToDashStyleConverter.cs:16]: It is actually being used in the new `TextBox` style, and I agree, the `ConvertBack()` should be implemented. Should probably be a relatively simple conversion anyway. I'll add that.

File: MaterialDesignThemes.Wpf/Converters/FloatingHintMarginConverter.cs [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+            || values[0] is not double prefixWidth
+            || values[1] is not Thickness prefixMargin
+            || values[2] is not double suffixWidth
+            || values[3] is not Thickness suffixMargin
+            || values[4] is not PrefixSuffixVisibility prefixVisibility
+            || values[5] is not PrefixSuffixVisibility suffixVisibility)
+        {
+            return new Thickness(0);
+        }
+        return new Thickness(GetLeftMargin(), 0, GetRightMargin(), 0);
```
[the remark this answers is not in the document]
Comment (nicolaihenriksen) [reply MaterialDesignThemes.Wpf/Converters/FloatingHintMarginConverter.cs:21]: Ooohh! I recalled having some issues with some of this syntax due to the `net462` TFM. I'll see if I can switch to list pattern matching.

File: MaterialDesignThemes.Wpf/Themes/MaterialDesignTheme.SmartHint.xaml [hunk 3]
```diff
@@ -182,14 +187,22 @@
...
+                        <MultiBinding Converter="{StaticResource FloatingHintScaleTransformConverter}">
                           <Binding ElementName="ScaleHost" Path="Scale" />
                           <Binding Path="FloatingScale" RelativeSource="{RelativeSource TemplatedParent}" />
                           <Binding Source="{StaticResource NoContentFloatingScale}" />
-                          <Binding Path="FloatingOffset" RelativeSource="{RelativeSource TemplatedParent}" />
-                          <Binding Path="InitialVerticalOffset" RelativeSource="{RelativeSource TemplatedParent}" />
                         </MultiBinding>
                       </ContentControl.RenderTransform>
+                      <Grid Background="{TemplateBinding wpf:HintAssist.Background}" IsHitTestVisible="False">
+                        <ContentControl Content="{TemplateBinding Hint}" IsHitTestVisible="False" IsTabStop="False">
```
[the remark this answers is not in the document]
Comment (nicolaihenriksen) [reply MaterialDesignThemes.Wpf/Themes/MaterialDesignTheme.SmartHint.xaml:197]: Not needed 👍

File: MaterialDesignThemes.Wpf/Converters/FloatingHintMarginConverter.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (nicolaihenriksen) [reply MaterialDesignThemes.Wpf/Converters/FloatingHintMarginConverter.cs]: Yea so it complains with this which is what I seemed to recall. Abandon NETFramework support? 😛 

image

File: MaterialDesignThemes.Wpf/Converters/BooleanToDashStyleConverter.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (nicolaihenriksen) [reply MaterialDesignThemes.Wpf/Converters/BooleanToDashStyleConverter.cs]: Pushed an implementation of `ConvertBack()`

File: MaterialDesignThemes.UITests/WPF/TextFieldDefaultHeightTests.cs [hunk 4]
```diff
@@ -30,22 +29,38 @@ public async Task SameHeightWithDefaultStyle()
...
-    <TextBox materialDesign:HintAssist.IsFloating=""True"" />
-    <PasswordBox materialDesign:HintAssist.IsFloating=""True"" />
-    <ComboBox materialDesign:HintAssist.IsFloating=""True"" />
-    <DatePicker materialDesign:HintAssist.IsFloating=""True"" />
-    <materialDesign:TimePicker materialDesign:HintAssist.IsFloating=""True"" />
+    <TextBox />
 </StackPanel>");
 
         var height = await GetHeight(stackPanel, "TextBox");
+        //Assert.Equal(height, await GetHeight(stackPanel, "ComboBox"), Precision);
```
Comment (nicolaihenriksen) [inline MaterialDesignThemes.UITests/WPF/TextFieldDefaultHeightTests.cs:43]: Comment left here simply to indicate what is described below:

In the other PR's adopting the new `SmartHint` you will see controls move from the original test method into the `Xxx_PostSmartHintRefactor()` method (one control per PR). An example is the ComboBox, where you can see this happen:
https://github.com/MaterialDesignInXAML/MaterialDesignInXamlToolkit/pull/3515/files#diff-e858ebe6ad71bec638332cb3f89c65f1570e5a3bcf92985e967c1ccb74f094d3

Once all styles have adopted the new approach, we can collapse into a single test with the original name again.

**Review by Keboo — APPROVED** [review]

**Pull request merged**